File: 剑指offer/12.矩阵中的路径.py
from typing import List


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        x_directions = [0, 0, -1, 1]
        y_directions = [-1, 1, 0, 0]

        n = len(board)
        if n == 0:
            return False
        m = len(board[0])

        def dfs(i, j, k):
            if k == len(word):
                return True

            if not 0 <= i < n or not 0 <= j < m or board[i][j] != word[k]:
                return False

            board[i][j] = ''
            res = False
            for q in range(4):
                if dfs(i + x_directions[q], j + y_directions[q], k + 1):
                    res = True
                    break
            board[i][j] = word[k]

            return res

        for i in range(len(board)):
            for j in range(len(board[0])):
                if dfs(i, j, 0):
                    return True

        return False

File: 剑指offer/12.矩阵中的路径.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    int m = 0;
    int n = 0;

    bool dfs(vector<vector<char>> board, const string word, int i, int j, int k)
    {
        if (i < 0 || i >= m || j < 0 || j >= n || board[i][j] != word[k]) {
            return false;
        }

        if (k == word.size() - 1) {
            return true;
        }

        board[i][j] = '\0';

        bool res = dfs(board, word, i, j + 1, k + 1) || dfs(board, word, i, j - 1, k + 1) || dfs(board, word, i + 1, j, k + 1) || dfs(board, word, i - 1, j, k + 1);

        board[i][j] = word[k];

        return res;
    }

    bool exist(vector<vector<char>>& board, string word)
    {
        m = board.size();
        n = board[0].size();

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (dfs(board, word, i, j, 0)) {
                    return true;
                }
            }
        }
        return false;
    }
};